File: wrench-sample/src/main/java/com/example/wrench/di/ViewModelModule.java
package com.example.wrench.di;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import com.example.wrench.WrenchSampleViewModel;
import com.example.wrench.livedataprefs.LiveDataPreferencesFragmentViewModel;
import com.example.wrench.wrenchprefs.WrenchPreferencesFragmentViewModel;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(WrenchSampleViewModel.class)
    abstract ViewModel bindSampleViewModel(WrenchSampleViewModel sampleViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(LiveDataPreferencesFragmentViewModel.class)
    abstract ViewModel bindLiveDataPreferencesFragmentViewModel(LiveDataPreferencesFragmentViewModel liveDataPreferencesFragmentViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(WrenchPreferencesFragmentViewModel.class)
    abstract ViewModel bindWrenchPreferencesFragmentViewModel(WrenchPreferencesFragmentViewModel wrenchPreferencesFragmentViewModel);

    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(ViewModelFactory factory);
}

File: wrench-sample/src/main/java/com/example/wrench/di/ViewModelModule.kt
package com.example.wrench.di

import android.app.Application
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider

import com.example.wrench.livedataprefs.LiveDataPreferencesFragmentViewModel
import com.example.wrench.service.WrenchServiceFragmentViewModel
import com.example.wrench.wrenchprefs.WrenchPreferencesFragmentViewModel

import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoMap

@Module
internal abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(LiveDataPreferencesFragmentViewModel::class)
    internal abstract fun bindLiveDataPreferencesFragmentViewModel(liveDataPreferencesFragmentViewModel: LiveDataPreferencesFragmentViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(WrenchPreferencesFragmentViewModel::class)
    internal abstract fun bindWrenchPreferencesFragmentViewModel(wrenchPreferencesFragmentViewModel: WrenchPreferencesFragmentViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(WrenchServiceFragmentViewModel::class)
    internal abstract fun bindWrenchServiceFragmentViewModel(wrenchServiceFragmentViewModel: WrenchServiceFragmentViewModel): ViewModel


    @Binds
    internal abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory
}
